File: generated-code/csharp/example/TcpReadWrite/OneOf.cs
namespace TransGenTest
{
    /// <summary>
    /// Oneof example
    /// </summary>
    public abstract class OneOf
    {
        /// <summary> Write OneOf to writer </summary>
        public abstract void WriteTo(System.IO.BinaryWriter writer);

        /// <summary> Read OneOf from reader </summary>
        public static OneOf ReadFrom(System.IO.BinaryReader reader)
        {
            switch (reader.ReadInt32())
            {
                case OptionOne.TAG:
                    return OptionOne.ReadFrom(reader);
                case OptionTwo.TAG:
                    return OptionTwo.ReadFrom(reader);
                default:
                    throw new System.Exception("Unexpected tag value");
            }
        }

        /// <summary>
        /// First option
        /// </summary>
        public class OptionOne : OneOf
        {
            public const int TAG = 0;
        
            /// <summary>
            /// List of integers
            /// </summary>
            public int[] VecInt { get; set; }
            /// <summary>
            /// Long integer
            /// </summary>
            public long LongInt { get; set; }
        
            public OptionOne() { }
        
            public OptionOne(int[] vecInt, long longInt)
            {
                this.VecInt = vecInt;
                this.LongInt = longInt;
            }
        
            /// <summary> Read OptionOne from reader </summary>
            public static new OptionOne ReadFrom(System.IO.BinaryReader reader)
            {
                var result = new OptionOne();
                result.VecInt = new int[reader.ReadInt32()];
                for (int vecIntIndex = 0; vecIntIndex < result.VecInt.Length; vecIntIndex++)
                {
                    result.VecInt[vecIntIndex] = reader.ReadInt32();
                }
                result.LongInt = reader.ReadInt64();
                return result;
            }
        
            /// <summary> Write OptionOne to writer </summary>
            public override void WriteTo(System.IO.BinaryWriter writer)
            {
                writer.Write(TAG);
                writer.Write(VecInt.Length);
                foreach (var vecIntElement in VecInt)
                {
                    writer.Write(vecIntElement);
                }
                writer.Write(LongInt);
            }
        
            /// <summary> Get string representation of OptionOne </summary>
            public override string ToString() {
                string stringResult = "OptionOne { ";
                stringResult += "VecInt: ";
                stringResult += "[ ";
                int vecIntIndex = 0;
                foreach (var vecIntElement in VecInt)
                {
                    if (vecIntIndex != 0) {
                        stringResult += ", ";
                    }
                    stringResult += vecIntElement.ToString();
                    vecIntIndex++;
                }
                stringResult += " ]";
                stringResult += ", ";
                stringResult += "LongInt: ";
                stringResult += LongInt.ToString();
                stringResult += " }";
                return stringResult;
            }
        }

        /// <summary>
        /// Second option
        /// </summary>
        public class OptionTwo : OneOf
        {
            public const int TAG = 1;
        
            /// <summary>
            /// usize
            /// </summary>
            public int Value { get; set; }
        
            public OptionTwo() { }
        
            public OptionTwo(int value)
            {
                this.Value = value;
            }
        
            /// <summary> Read OptionTwo from reader </summary>
            public static new OptionTwo ReadFrom(System.IO.BinaryReader reader)
            {
                var result = new OptionTwo();
                result.Value = reader.ReadInt32();
                return result;
            }
        
            /// <summary> Write OptionTwo to writer </summary>
            public override void WriteTo(System.IO.BinaryWriter writer)
            {
                writer.Write(TAG);
                writer.Write(Value);
            }
        
            /// <summary> Get string representation of OptionTwo </summary>
            public override string ToString() {
                string stringResult = "OptionTwo { ";
                stringResult += "Value: ";
                stringResult += Value.ToString();
                stringResult += " }";
                return stringResult;
            }
        }
    }
}

File: generated-code/fsharp/example/TcpReadWrite/OneOf.fs
#nowarn "0058"

namespace TransGenTest

/// First option
type OneOfOptionOne = {
    /// List of integers
    VecInt: int[];
    /// Long integer
    LongInt: int64;
} with

    /// Write OptionOne to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 0
        writer.Write this.VecInt.Length
        this.VecInt |> Array.iter (fun value ->
            writer.Write value )
        writer.Write this.LongInt

    /// Read OptionOne from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        VecInt = [|for _ in 1 .. reader.ReadInt32() do
                     yield reader.ReadInt32() |]
        LongInt = reader.ReadInt64()
    }

/// Second option
type OneOfOptionTwo = {
    /// usize
    Value: int;
} with

    /// Write OptionTwo to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 1
        writer.Write this.Value

    /// Read OptionTwo from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        Value = reader.ReadInt32()
    }

/// Oneof example
type OneOf =
    /// First option
    | OptionOne of OneOfOptionOne
    /// Second option
    | OptionTwo of OneOfOptionTwo
    with

    /// Write OneOf to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        match this with
            | OptionOne value -> value.writeTo writer
            | OptionTwo value -> value.writeTo writer

    /// Read OneOf from reader
    static member readFrom(reader: System.IO.BinaryReader) =
        match reader.ReadInt32() with
            | 0 -> OptionOne (OneOfOptionOne.readFrom reader)
            | 1 -> OptionTwo (OneOfOptionTwo.readFrom reader)
            | x -> failwith (sprintf "Unexpected tag %d" x)